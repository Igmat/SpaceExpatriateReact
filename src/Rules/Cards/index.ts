import {
    action,
    computed,
    makeObservable,
    observable,
  } from "mobx";
  import { HandModel } from "../HandModel";
  import { TableModel } from "../TableModel";
 import { ICardPlace } from "./ICardPlace";
  import { DeckModel } from "../DeckModel";
  
  export class CardsMethods {
    public place?: "hand" | "deck" | "table" = undefined;
    constructor() {
      makeObservable(this, {
        place: observable,
        isInHand: computed,
        isInDeck: computed,
        isOnTable: computed,
        move: action.bound,
      });
    }
    public get isInHand() {
      return this.place === "hand";
    }
  
    public get isInDeck() {
      return this.place === "deck";
    }
    public get isOnTable() {
      return this.place === "table";
    }
  
    public move(from: ICardPlace, to: ICardPlace) {
      to.placeCard(from.takeCard(this as any));
      //console.log(to)
      if (to instanceof TableModel) {
        this.place = "table";
      }
      if (to instanceof DeckModel) {
        this.place = "deck";
      }
      if (to instanceof HandModel) {
        this.place = "hand";
      }
  
     // console.log(this.place);
    }

  }