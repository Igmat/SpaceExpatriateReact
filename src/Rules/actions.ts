type modalActionType = {
    [key: string]: string[]
}

export const modalAction: modalActionType = {
    "terraforming": ["Build Colony", "Do nothing"],
}
    