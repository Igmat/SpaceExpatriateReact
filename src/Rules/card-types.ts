export type ResourcePrimitive = 'fuel' | 'minerals' | 'biotic materials' | 'machinery' | 'nanotechnologies' | 'dark matter';

export const isResourcePrimitive = (option: string ): option is ResourcePrimitive => 
    (['fuel', 'minerals', 'biotic materials', 'machinery', 'nanotechnologies', 'dark matter']).includes(option)

export const isCardType = (option: string): option is CardType =>
    (['delivery', 'engineering', 'terraforming', 'military']).includes(option)

export type Resource = ResourcePrimitive | ResourcePrimitive[];

export type CardDefinition = DeliveryCard | EngineeringCard | TerraformingCard | MilitaryCard;

export type CardType = 'delivery' | 'engineering' | 'terraforming' | 'military'

export interface DeliveryCard {
    id: number,
    type: 'delivery',
    resources: /*ResourcePrimitive | */ResourcePrimitive[],
   // points: number
}

export interface EngineeringCard {
    id: number,
    type: 'engineering',
    connection: 'start' | 'continue' | 'end',
    entryPoint?: ResourcePrimitive,
    exitPoint?: ResourcePrimitive[],
    points?: number,
    name: string
}

export interface TerraformingCard {
    name: string,
    id: number,
    type: 'terraforming',
    resources: Resource[],
    points: number
}

export interface MilitaryCard {
    id: number,
    type: 'military',
    weapon: 'orbital' | 'intelligence' | 'fighters' | 'spaceborne',
    name: string,
   // points: number
}