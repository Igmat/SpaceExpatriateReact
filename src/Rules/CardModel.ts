import { makeAutoObservable } from "mobx";
import { CardType } from "./card-types";

export class CardModel {
  constructor(public id: number, public type: CardType) {
    makeAutoObservable(this);
  }
}
