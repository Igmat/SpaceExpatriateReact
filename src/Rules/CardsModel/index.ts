import { CardType, ColonyCard } from "../card-types";
import { makeAutoObservable } from "mobx";
import { CardDefinition } from "../card-types";
import { GeneralCardDefinition } from "../CardDefinitions/createCards";
export class CardsModel {

    constructor (
        id: number,
        type: GeneralCardDefinition
    ) {
       // makeAutoObservable(this);

}

test(){
console.log('!!!!!!!!!!!!')
}

}