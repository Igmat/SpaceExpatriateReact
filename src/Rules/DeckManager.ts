import { makeAutoObservable } from "mobx";
import { deliveryCards } from "./CardDefinitions/delivery";
import { engineeringCards } from "./CardDefinitions/engineering";
import { militaryCards } from "./CardDefinitions/military";
import { terraformingCards } from "./CardDefinitions/terraforming";
import { DeckModel } from "./DeckModel";
import { CardDefinition, ColonyCard } from "./card-types";

export class DeckManager {
  constructor(private readonly gameId: string) {
    makeAutoObservable(this);
  }

  delivery = new DeckModel("delivery", deliveryCards, this.gameId);
  engineering = new DeckModel("engineering", engineeringCards, this.gameId);
  terraforming = new DeckModel("terraforming", terraformingCards, this.gameId);
  military = new DeckModel("military", militaryCards, this.gameId);

  dropCards = (...cards: Exclude<CardDefinition, ColonyCard>[]) => {
    cards.forEach((card) => this[card.type].dropCards(card.id));
  };
  get dropLength() {
    return (
      this.delivery.restCount +
      this.engineering.restCount +
      this.terraforming.restCount +
      this.military.restCount
    );
  }

}
