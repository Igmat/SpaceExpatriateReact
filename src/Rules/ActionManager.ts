import { makeAutoObservable } from "mobx";
import { DeckManager } from "./DeckManager";
import { CardType, GeneralCard } from "./card-types";
import { TableModel } from "./TableModel";
import { RoundManager } from "./RoundManager";
import { HandModel } from "./HandModel";
import { ResourcesModel } from "./ResourcesModel";
import { ActionManager as EAM } from "./Engineering";
import { ActionManager as TAM } from "./Terraforming";
import { ActionManager as DAM } from "./Delivery";
import { ActionManager as MAM } from "./Military";
import { makeAutoSavable } from "../Utils/makeAutoSavable";
import { ColonyManager } from "./Colony/ColonyManager";
import { ColonyDeckModel } from "./Colony/ColonyDeckModel";
import { ModalManager } from "./ModalManager";
import { GameState } from ".";

export class ActionManager {
  constructor(
    private readonly gameState: GameState,
    private readonly decks: DeckManager,
    private readonly table: TableModel,
    private readonly round: RoundManager,
    private readonly hand: HandModel,
    private readonly resources: ResourcesModel,
    private readonly gameId: string,
    private readonly colony: ColonyManager,
    private readonly colonyDeck: ColonyDeckModel,
    private readonly modal: ModalManager,
  ) {
    makeAutoObservable(this);
    makeAutoSavable(this, gameId, `action`, [`activeAction`],this.gameState.saveCondition);
  }

  private managers = {
    engineering: new EAM(
      this.round,
      this.table,
      this.decks,
      this.hand,
      this.gameId
    ),
    terraforming: new TAM(
      this.round,
      this.table,
      this.decks,
      this.gameId,
      this.colony,
      this.colonyDeck,
      this.resources,
      this.modal,
      this.hand
    ),
    delivery: new DAM(
      this.gameState,
      this.table,
      this.round,
      this.hand,
      this.resources,
      this.decks,
      this.modal,
      this.gameId
    ),
    military: new MAM(this.round, this.hand, this.decks, this.modal),
  };

  activeAction?: CardType;

  get currentManager() {
    return this.activeAction && this.managers[this.activeAction];
  }

  perform = async (card?: GeneralCard) => {
    if (!card) return;
    if (this.round.phase !== "active") return;
    this.activeAction = card.type;
    card.move(this.table.columns[card.type])

    if (this.round.current < 5) {
      this.nextRound();
      return;
    }

    this.round.phase = card.type;
    await this.colony.triggers.before(this.activeAction);
    await this.currentManager?.perform(card);
    await this.colony.triggers.afterSelect(this.activeAction);
  };

  nextRound = async () => {
    this.activeAction && (await this.colony.triggers.after(this.activeAction));
    this.colony.cancelActiveEffects();
    this.round.next();
    this.activeAction = undefined;
  };

  confirm = async () => {
    await this.currentManager?.confirm();
    this.currentManager?.isEnded && await this.nextRound();
  };

  activateDeck = async (type: CardType) => {
    this.currentManager?.activateDeck(type);
    this.currentManager?.isEnded && await this.nextRound();
  };

  activateCard = async (card: number) => {
    this.currentManager?.activateCard(card);
    this.currentManager?.isEnded && await this.nextRound();
  };

  activateColonyCard = async (card: number) => {
    this.currentManager?.activateColonyCard(card);
    this.currentManager?.isEnded && await this.nextRound();
  };

  activateCardOnTable = async (card: GeneralCard) => {
    await this.currentManager?.activateCardOnTable(card);
    this.currentManager?.isEnded && await this.nextRound();
  };

  reset = async () => await this.currentManager?.reset();

  get isDisabled(): (card:GeneralCard) => boolean {
    return (card:GeneralCard) => {
      if (!this.activeAction) return !this.decks.isInDeck(card);
      return this.managers[this.activeAction].isDisabled(card);
    };
  }

  get isDisabledDeck(): (type: CardType) => boolean {
    return (type: CardType) => {
      if (!this.activeAction) return true;
      return this.managers[this.activeAction].isDisabledDeck(type);
    };
  }
}
