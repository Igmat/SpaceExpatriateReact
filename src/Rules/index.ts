import { HandModel } from "../Rules/HandModel";
import { DeckManager } from "./DeckManager";
import { TableModel } from "./TableModel";
import { makeAutoObservable } from "mobx";
import { RoundManager } from "./RoundManager";
import { ActionManager } from "./ActionManager";
import { ResourcesModel } from "./ResourcesModel";
import { createContext, useContext } from "react";
import { ColonyManager } from "./Colony/ColonyManager";
import { ColonyDeckModel } from "./Colony/ColonyDeckModel";
import { colonyCards } from "./Colony/colony-cards";

export class GameState {
  constructor(public readonly gameId: string = "") {
    makeAutoObservable(this);
  }

  hand = new HandModel(this.gameId);
  decks = new DeckManager(this.gameId);
  colonyDeck = new ColonyDeckModel(colonyCards, this.gameId);
  table = new TableModel(this.gameId);
  round = new RoundManager(this.decks, this.hand, this.colonyDeck, this.gameId);
  resources = new ResourcesModel(this.table, this.round, this.gameId);
  colony = new ColonyManager(this, this.gameId, this.table, this.round, this.resources);
  action = new ActionManager(this.decks, this.table, this.round, this.hand, this.resources, this.gameId, this.colony, this.colonyDeck);
}

const gameStateContext = createContext(new GameState())
export const { Provider } = gameStateContext;
export const useGameState = () => useContext(gameStateContext);
