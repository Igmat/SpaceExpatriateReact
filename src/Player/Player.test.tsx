import { Player } from ".";
import renderer from 'react-test-renderer';
import { GameState, Provider } from "../Rules";

const originalRandom = Math.random;

beforeEach(() => Math.random = () => 1 / 2);

afterAll(() => Math.random = originalRandom);

test('renders player', () => {

  const mockedGameState = new GameState()

  const tree = renderer.create(
    <Provider value={mockedGameState}>
      <Player />
    </Provider>
  ).toJSON();
  expect(tree).toMatchSnapshot();
});
