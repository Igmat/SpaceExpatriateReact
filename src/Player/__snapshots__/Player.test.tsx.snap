// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders player 1`] = `
Array [
  <div
    className="container"
  >
    <div
      className="cardsContainer"
    >
      <div
        className="card"
      >
        <div
          className="mutateAction"
        >
          terraforming
        </div>
        <div
          className="benefit"
        >
          Rerum voluptatum optio voluptatem a quae saepe. Cumque quibusdam placeat corporis.
        </div>
        <p
          className="points"
        >
          2
        </p>
        <p
          className="colony"
        >
          COLONY
        </p>
        <p
          className="name"
        >
          tra la la la la la
        </p>
      </div>
    </div>
    <div
      className="cardsContainer"
    />
    <div
      className="cardsContainer"
    />
    <div
      className="cardsContainer"
    />
    <div
      className="cardsContainer"
    />
    <div
      className="round"
    >
      Round: 1
    </div>
  </div>,
  <div
    className="container"
  >
    <div
      className="delivery card   "
      onClick={[Function]}
    >
      <div>
        <div>
          <div
            className="nanotechnologies resource"
          >
            <div
              className="resName"
            >
               
              nanotechnologies
            </div>
          </div>
           
          <div />
        </div>
      </div>
    </div>
    <div
      className="engineering card   "
      onClick={[Function]}
    >
      <div
        className="start"
      >
        <div>
          start
        </div>
        <div
          className="entryPoint"
        >
          <div>
            <div>
              <div
                className="biotic-materials resource"
              >
                <div
                  className="resName"
                >
                   
                  biotic-materials
                </div>
              </div>
               
              <div />
            </div>
          </div>
        </div>
        <div
          className="exitPoint"
        >
          <div>
            1
          </div>
        </div>
      </div>
    </div>
    <div
      className="military card   "
      onClick={[Function]}
    >
      <div
        className="intelligence"
      >
        Intelligence
      </div>
    </div>
    <div
      className="terraforming card   "
      onClick={[Function]}
    >
      <div
        className="infoWrapper"
      >
        <div
          className="resourceWrapper"
        >
          <div>
            <div>
              <div
                className="nanotechnologies resource"
              >
                <div
                  className="resName"
                >
                   
                  nanotechnologies
                </div>
              </div>
               
              <div />
            </div>
          </div>
          <div>
            <div>
              <div
                className="nanotechnologies resource"
              >
                <div
                  className="resName"
                >
                   
                  nanotechnologies
                </div>
              </div>
               
              <div />
            </div>
          </div>
        </div>
        <div
          className="points"
        >
          4
        </div>
        <p
          className="cardName"
        >
          NANOREPLICATOR
        </p>
      </div>
    </div>
  </div>,
  <div
    className="container"
  >
    <div
      className="playerResources"
    >
      <div
        className="biotic"
      >
        0
        <div
          className="notification"
        />
      </div>
      <div
        className="fuel"
      >
        0
        <div
          className="notification"
        />
      </div>
      <div
        className="machinery"
      >
        0
        <div
          className="notification"
        />
      </div>
      <div
        className="minerals"
      >
        0
        <div
          className="notification"
        />
      </div>
      <div
        className="nanotechnologies"
      >
        0
        <div
          className="notification"
        />
      </div>
      <div
        className="dark"
      >
        0
        <div
          className="notification"
        />
      </div>
      <div
        className="energy"
      >
        Energy: 
        0
      </div>
      <div
        className="points"
      >
        TPs
        0
      </div>
    </div>
  </div>,
]
`;
